### ISSUE TITLE
Dynamic Library linking and Object Database

### ISSUE THREAD

**Issue by FoamScience** [su2code/SU2#1058]
Hi everyone,

First of all, I want to thank you for this marvelous project you're running! Well done!
But I have some suggestions about the "building" and design philosophy of SU2

This (programming-related) "feature request" is all about making SU2 even more extendable

## Is my feature request related to a problem?
Maybe! It's not pleasant when I'm forced to link libraries statically. I think it hinders the potential of the C++ language and makes compilation time unnecessarily longer.

For example, in [SU2_CFD.cpp](https://github.com/su2code/SU2/blob/master/SU2_CFD/src/SU2_CFD.cpp#L117) the program decides what to do with an if statement and **all drivers need to be linked at compile-time**. What if I want to write my own driver? I need to include my headers in "SU2_CFD.hpp" and rebuild the whole thing again!!!

## The solution I would like
Why not use a [Factory With Self-Registering Classes](https://www.bfilipek.com/2018/02/factory-selfregister.html) instead. Ideally, the parent class (CDriver for example) would have
a static ::new() (or ::create() ) method which will create the appropriate type (child classes) based on **user input** (aka, from config file).
Of course, this way, the table of registered classes (drivers in my example) could be extended if the user requests to
dynamically load his own library (assuming that the library contains a child class that registers itself to the table).

There is another vital related issue: Incorporating SU2 code in user-developed code is the main concern of this feature request but we also need a mechanism for getting references to all created objects. In the shared separate library, the user expects to
use already created CVariables, ... etc; how can he get those? An object database that keeps references to most involved objects is needed.

**This roughly describes what I think is needed to be done. I'm willing to help and even lead the development of this feature if it gets enough approval**.

### Alternatives?
Can't think of anything else! This feature will enable devs to effortlessly extend the software and it even facilitates the building of User Defined Functions in the future as it'll become an important feature when the project gains enough popularity.

The only other option is to leave all this behind and continue developing following the current design until the project reaches a state where it will become very hard to go back and implement this feature.

**Additional context**
1. The design I described above doesn't take into account C++ templates as I don't see them used that frequently (Is that a thing?)

2. One last thing: Please, Please, we have to stop using raw pointers like that. I ran valgrind on the "Inviscid Bump in a Channel" case and discovered that 1Megs got definitely lost (leaked). Of course it will, if raw pointers get used in that way. I fully understand that many of the skilled programmers here come from C/Fortran background but please refrain from using raw pointers everywhere
and use C++ memory management objects (weak_ptr, unique_ptr, shared_ptr ... etc) instead. Make it a coding style if you will.

**Comment by pcarruscag** [comment]
Hi,
I like this idea of self registering classes for our factories, although the instantiation logic is seldom as linear as converting a name to a type... Maybe you could refactor just one of our factories to give us a better idea of what it could look like?

This issue of "modularizing" pieces of the code is central to the refactoring of the optimization framework, which will require significant chunks of the code to be exposed to python, so by all means get involved in that (join #refactor_optimization on SU2 Slack and the developer meetings, Wednesdays at 4pm CET).

Personally I think this flavour of extensibility is small fish compared to other structural problems we are currently addressing (duplication, code bloat, performance, robustness, etc.) and in my opinion addressing those issues will make it easier to extend the code in this manner, not harder.
Also from the pure C++ side I see no problem with having to modify the code... In so doing the new code is automatically under the same open source license, which to me is as good thing.

A couple of practical considerations:
- About your examples, we want fewer drivers, not more. The methods of CVariable, CPoint, and co. need to be inline if the code is to have any semblance of performance.
- I think the majority of the code should still be statically linked, even if some parts of it can be extended by means of custom dynamic libraries, because that will just be less problematic for folks who build from source but are not developers.

> The design I described above doesn't take into account C++ templates as I don't see them used that frequently (Is that a thing?)

It is becoming a thing yes absolutely, although not at the very highest levels of abstraction.

> Please, Please, we have to stop using raw pointers like that

It's being done... It takes time and not many people are comfortable doing this kind of refactoring (consider it an invitation :) ).

**Comment by FoamScience** [comment]
Thanks for the quick reply.

- I will try to craft a working example with "CSolverFactory" and a couple of existing solvers
- It seems you have your hands full (didn't see the 5 projects before :) ). Naturally, I'll help whenever I can.

**Comment by FoamScience** [comment]
OK, I think I did some progress regarding this issue.

After you read this, please [Compare](https://github.com/FoamScience/SU2/compare/master...FoamScience:feature_self_reg_classes) the master and the feature branches.

## Design Idea

- We need to register a child class to a map stored in a factory class; obviously, well before construction time.
- The static variables to the rescue: Associate a static variable to the class and call the registration function when initializing the static variable. done!

## Work Status

> I'm using "Inviscid Bump in the Channel" case for testing.

1. I had to make some changes to the building options
   - Upgrading to "C++14" allowing me to use make_unique; a personal preference.
   - "--whole-archive" GCC compiler option is necessary in order to pull and force-initialize all symbols from statically-linked library. This is compiler-specific and may bloat the result binary with unnecessary symbols. But it's the quickest way.
2. I tried to make changes so the interaction with the code is minimal so I went and took advantage of **CSolverFactory::CreateSubSolver()** to create a pointer to self-registered **CEulerSolver**.
3. By the way, I thought the solver would be created only once. Little I knew!
4. Anyway, The registration works as expected now. If you change "EULER" in [this line](https://github.com/FoamScience/SU2/blob/feature_self_reg_classes/SU2_CFD/src/solvers/CSolverFactory.cpp#L226) to "SOMETHINGELSE", you get:
```
-------------------- Solver Preprocessing ( Zone 0 ) --------------------
<---------- Solver SOMETHINGELSE was not found.
<---------- Available Solvers: 
EULER 
``` 
and the program exists.
5. However, after returning the created solver out of **CSolverFactory::CreateSubSolver()** some strange things happen:
```
-------------------- Solver Preprocessing ( Zone 0 ) --------------------
<----------- Making CEulerSolver ...
...... Some other output
Initialize Jacobian structure (Euler). MG level: 0.
<----------- Preprocessing Created solver ...
<----------- Making CEulerSolver ...
Initialize Jacobian structure (Euler). MG level: 1.
<----------- Preprocessing Created solver ...
<----------- Making CEulerSolver ...
Initialize Jacobian structure (Euler). MG level: 1.
<----------- Preprocessing Created solver ...
<----------- Making CEulerSolver ...
Initialize Jacobian structure (Euler). MG level: 1.
<----------- Preprocessing Created solver ...
<----------- Making CEulerSolver ...
.... And then
Elwardi-PC:04092] *** Process received signal ***
[Elwardi-PC:04092] Signal: Segmentation fault (11)
[Elwardi-PC:04092] Signal code: Address not mapped (1)
[Elwardi-PC:04092] Failing at address: (nil)
[Elwardi-PC:04092] [ 0] /lib/x86_64-linux-gnu/libc.so.6(+0x3efd0)[0x7f6f2de1dfd0]
[Elwardi-PC:04092] *** End of error message ***
Segmentation fault
```
So there is probably something else I didn't pay attention to.

## Immediate improvements

Forcing the dev to create his own "is_registered" static variable and initializing it himself can be eliminated by by putting the variable in a templated class (notice that the variable initialization only depends on the type it belongs to and the factory class) which the solver can inherit from.

## My question

Any idea on what is causing the seg fault. I'm pretty sure it's because I'm returning a nullptr somewhere though.

**Comment by pcarruscag** [comment]
Looks very interesting, maybe you can open a draft pull request so we can comment directly on the code.
At first glace your segfault is because the unique pointer goes out of scope and so the return from the factory is a dangling pointer.

**Comment by pcarruscag** [comment]
Just another note, portability is one of the cornerstones of SU2, so please try to work within C++11 boundaries, a lot of C++14 features are syntax sugar on top of C++11 capabilities, you can just adapt the example implementations from the online c++ documentation (no boost :) ).
Also compiler specific flags will be a bit of a deal breaker if they don't have support from the major vendors (gcc, clang, intel, ms).

**Comment by FoamScience** [comment]
I went ahead and opened a draft PR, Back to C++11 (although now I'm using raw pointers :( ) and removed the dependency on compiler-specific flags (even if --whole-archive is supported by most compilers).

**Comment by economon** [comment]
Just in case it is helpful, the Abseil library from Google is a nice way to use some C++14 and later features with only a C++11 requirement. We could consider integrating if it makes life much easier...

https://abseil.io

**Comment by FoamScience** [comment]
> Just in case it is helpful, the Abseil library from Google is a nice way to use some C++14 and later features with only a C++11 requirement. We could consider integrating if it makes life much easier...
> 
> https://abseil.io

Thanks for pointing me to this library :) There was a project that was in dire need for such libraries.

Well I'm still a newbie discovering the code. From what I can see, the whole code base is C-oriented (maybe even originating from Fortran??). But I would advise against integrating it (Maybe it's a bit of an overkill for the current state of SU2) because most devs won't actually use the features from c++17 (myself included) and we should be able to deal with the lack of make_unique and variable templates in c++11 :smile:.

With that said, if you're planning on using more c++ features while rebuilding/extending the code base, it might be a very good idea to integrate it.

**Comment by bigfooted** [comment]
For what it's worth, it's very easy to dynamically link files to SU2, for instance if you want to add custom boundary conditions or source terms. Just put the default definition of the function in a dynamically linked library and when you want to use a different definition of this function, just point your library path to your custom library in e.g. your testcases directory. If this is all you want, you can do that without any additional libraries. This is more or less the same way as e.g. ANSYS Fluent does it.

**Comment by FoamScience** [comment]
> Just put the default definition of the function in a dynamically linked library and when you want to use a different definition of this function, just point your library path to your custom library in e.g. your testcases directory.

Could you elaborate more on this?? Let me explain a typical scenario I would go through (picking BCs as an example):
-  I should be able to construct my custom BC inheriting from one of the existing ones. Put this new BC_TYPE boundary condition in my dynamically-linked library.
- BUT no recompiling of SU2_CFD; the program should provide facilities to dynamically load libs at RUNTIME (e.g. list of custom ibs in case config file).
- Now, switching between my custom BC types and the standard ones only requires to edit the config file.

As I understand,
- Dynamically loading a library specified in the config file at runtime is easy. "dlopen" is there, and possibly more sophisticated libraries?
- The issue is that there is no way to "select" the new BC because it was not known at compile-time.
- Most of the code is built around the assumption of "knowing things" at compile time. So a series of if statements checking the setting in the config file was the way to go. If the setting in the config says EULER; The Euler BC gets picked up. But what if it says Custom_EULER?

I got distracted by some other stuff. But here is the thing:
- All BCs should inherit from a Base Class (Assuming BCs are classes to begin with)
- The selection mechanism will rely on the custom BC class registering itself to a Factory automatically.
- Through the magic of templates and macro expansion, it's possible to easily add the custom BC to the map of known BCs (Yet to be implemented).

If this interests you and if you think there is a decent chance of incorporating this design into the code, please let me know. I will give this a higher priority :)

**Comment by bigfooted** [comment]
> Could you elaborate more on this?? Let me explain a typical scenario I would go through (picking BCs as an example):
> 
>     * I should be able to construct my custom BC inheriting from one of the existing ones. Put this new BC_TYPE boundary condition in my dynamically-linked library.
> 
>     * BUT no recompiling of SU2_CFD; the program should provide facilities to dynamically load libs at RUNTIME (e.g. list of custom ibs in case config file).
> 
>     * Now, switching between my custom BC types and the standard ones only requires to edit the config file.

For example, the inlet BC of the SA turbulence model is defined in CTurbSASolver::BC_Inlet()
We can place this definition in a separate file bcinit.cpp and dynamically link this file so you have a libbcinit.so file. SU2 knows about libbcinit.so and BC_Inlet() at compile time. 
In your testcases directory, you can put a copy of bcinit.cpp and redefine the boundary condition BC_Inlet. If you change the boundary condition, you only compile bcinit.cpp in your testcase directory and create a libbcinit.so in your testcase directory. You then make sure that when SU2 is looking for dynamically linked libraries, it searches your testcase directory before it searches the SU2 installation directory.  

As you see, this can be used for redefining existing (at compile time) functions and modifying its content. I find it especially useful for creating custom initial conditions. If you want to create a completely new custom boundary condition BC_Custom(), it is probably easiest to predefine this structure and create a custom_bc option in the config file. In a CFD problem, there are only a limited number of possibilities (for types of boundary conditions, and systems of equations), so I'm not sure when a BC_Custom() is useful, or when the tipping point is where self-registering classes will become useful for its implementation. Recompiling the code completely takes maybe 2 minutes for me, so I never saw compiling times as a problem. However, this concept could reduce code duplication and improve readability and maintainability, so I'm curious to see the progress on this...

**Comment by FoamScience** [comment]
Ah, OK! Tricking the linker!

It works well until you get this idea of building a library on top of SU2, for example to simulate Reservoir Engineering problems (which I consider to be a subset of CFD and many other things).

Overwriting the default behavior of functions and classes becomes impractical for such cases :smile:.

I'll attempt to complete the work on this feature in a couple of days then.

**Comment by FoamScience** [comment]
@bigfooted @economon  and @pcarruscag 

I created separate repo to demonstrate what I'm after in the simplest way possible: [FoamScience/AutoRegistering-Cpp-Classes](https://github.com/FoamScience/AutoRegistering-Cpp-Classes/commits/master)

> If you're going to check the code, check CSolver first, then CEulerSolver ...

- Commit FoamScience/AutoRegistering-Cpp-Classes@20ca601 implements the concept for a single Factory.
  - (Make-believe) Standard solver classes are compiled to a library
  - A (Make-beleive) Custom solver class is compiled to its own shared library
  - By default, the make program knows only about CEulerSolver (from standard solvers lib)
  - But if you pass in a library name (eg. libCCustomSolvers.so), it gets loaded, symbols pulled and the registration map for CSolver gets updated with the new Solver automatically.

- But it would bloat the code if things went this way, so commit FoamScience/AutoRegistering-Cpp-Classes@380c052 isolates self-registration code into 3 macros to:
   -  Declare the registration map and associated members
   -  Define and initialize registration-related members on the base class
   -  Register a class

If you only care about the interface and not how the thing is implemented, commit 380c052 is your target.
I took care to document the important parts of the code, so, please, dig in and let me know what you think.

At this point, integrating this with SU2 (and at what parts, and when) is your call but If you decide to do so, l will help.
> And yes, I wrote my own build script so I can have direct control over the compiler and the linker.


**One last note: I used "dlopen", so you will have to compile the code on Linux I suppose.**
